### PR TITLE
IMPRO-296 Snow falling probability - Percentile to Probability

### PR DESCRIPTION
**Pull request by bayliffe**
A plugin to interpolate percentiles to intermediate values to extract probabilities.

There is a notebook relating to IMPRO-296 in improver_aux that includes this plugin with demonstrations of its use, and which outlines a preferred method for using it in the suite.

There are currently no tests to go with this plugin, these will be added after it has been looked at.

### CODE REVIEW

File: lib/improver/utilities/statistical_operations.py [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+        The orography heights are compared against the heights that
+        correspond with percentile values to find the band in which they
+        fall; this diagram hides the 2-dimensional variability of the snow
+        fall level. The percentile values are then interpolated to the
+        height of the point being considered. This constructs a
+        2-dimensional field of probabilities that snow will be falling at
+        each point in the orography field.
+    """
+
+    def __init__(self, reference_cube, output_name, inverse_ordering=False):
```
Comment (carolinejjones) [inline lib/improver/utilities/statistical_operations.py]: This plugin feels the wrong way around. The "reference_cube" is actually the latest data which will be updating each day and the cube is the static field of orography. It becomes more obvious further down when we base the new probability field on the "reference_cube".

The ability to set your own output_name is useful but I don't think it should be compulsory to supply this. It should default to "probability_of'+reference_cube.name() if not supplied.

File: lib/improver/utilities/statistical_operations.py [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+                The input cube to be used as a template.
+        Returns:
+            probability_cube (iris.cube.Cube):
+                The new cube with suitable metadata.
+        """
+        cube_format = next(cube.slices([cube.coord(axis='y'),
+                                        cube.coord(axis='x')]))
+        probabilities = cube_format.copy(data=np.full(cube_format.shape, 1.,
+                                                      dtype=float))
+        percentile_coordinate, = [coord.name() for coord in
```
Comment (carolinejjones) [inline lib/improver/utilities/statistical_operations.py]: You are looking for percentile coordinate in several places within this code and I know it is being looked for elsewhere (e.g. blending). It would probably make sense to add a function to the utilities code to do this search and use this same bit of code everywhere. Not that this is a large bit of code but if we change the way we define a percentile coordinate we would then only have to change the search in the utility rather than everywhere within the code.

File: lib/improver/utilities/statistical_operations.py [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+            cube (iris.cube.Cube):
+                The input cube to be used as a template.
+        Returns:
+            probability_cube (iris.cube.Cube):
+                The new cube with suitable metadata.
+        """
+        cube_format = next(cube.slices([cube.coord(axis='y'),
+                                        cube.coord(axis='x')]))
+        probabilities = cube_format.copy(data=np.full(cube_format.shape, 1.,
+                                                      dtype=float))
```
Comment (carolinejjones) [inline lib/improver/utilities/statistical_operations.py]: Data gets set everywhere to 1.0, would it be better to set this to np.nan or some other missing indicator in case the value never gets set properly?

File: lib/improver/utilities/statistical_operations.py [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+    def percentile_interpolation(self, cube, reference_cube):
+        """
+        Perform the percentile interpolation to construct the probability
+        field.
+
+        Args:
+            cube (iris.cube.Cube):
+                A cube of values, that effectively behave as thresholds, for
+                which it is desired to obtain probability values from a
+                percentiled reference cube.
```
Comment (carolinejjones) [inline lib/improver/utilities/statistical_operations.py]: reference_cube needs to be added to Args although I think these would be better named cube -> threshold_cube, reference_cube -> percentile_cube

File: lib/improver/utilities/statistical_operations.py [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+        array_shape = tuple(array_shape)
+        percentile_bounds = np.full(array_shape, -1, dtype=float)
+        height_bounds = np.full(array_shape, -1001., dtype=float)
+        height_bounds[1] = -1.
+
+        percentile_coordinate, = [coord for coord in reference_cube.coords()
+                                  if 'percentiles' in coord.name()]
+        percentiles = percentile_coordinate.points
+
+        for index, pslice in enumerate(reference_cube.slices_over(
```
Comment (carolinejjones) [inline lib/improver/utilities/statistical_operations.py]: Could np.interp be used here rarely than looping through all the percentiles?

File: lib/improver/utilities/statistical_operations.py [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+                percentile_coordinate)):
+            indices = (cube.data < pslice.data if self.inverse_ordering else
+                       cube.data > pslice.data)
+            percentile_bounds[0, indices] = percentiles[index]
+            height_bounds[0, indices] = pslice.data[indices]
+            try:
+                # Usual behaviour where the orography falls between heights
+                # corresponding to percentiles.
+                percentile_bounds[1, indices] = percentiles[index + 1]
```
Comment (carolinejjones) [inline lib/improver/utilities/statistical_operations.py]: What happens if several percentiles have the same value

File: lib/improver/utilities/statistical_operations.py [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+            output_cubes (iris.cube.Cube):
+                A cube of probabilities obtained by interpolating the
+                percentile values.
+        """
+        try:
+            self.reference_cube.coord_dims('realization')
+        except CoordinateNotFoundError:
+            cube_slices = [self.reference_cube]
+        else:
+            cube_slices = self.reference_cube.slices_over('realization')
```
Comment (carolinejjones) [inline lib/improver/utilities/statistical_operations.py]: Why only slicing over realization? What if the percentiles have other coords e.g. Time?

File: lib/improver/utilities/statistical_operations.py (hunk 6 above)
Comment (cgsandford) [reply lib/improver/utilities/statistical_operations.py]: As a coordinate, I think the percentile value (ie on the 0-100 spectrum) should not be multiply defined?  If the height value is the same for several different percentile levels, this is caught by ignoring the divide-by-zero in calculating the interpolant.  The output probability then relates to the lowest of the possible percentiles - eg if a height of 4.0 km corresponded to the 0th and 10th percentiles, the code as written would assign that a probability of "0" (0.1 if across the 10th and 20th percentiles, etc).  But there are no problems with singularity or NANs.

I've added a unit test for this behaviour.  We may need to think about whether this is actually what we want, and what the ideal behaviour would look like.

File: lib/improver/utilities/statistical_operations.py (hunk 7 above)
Comment (gavinevans) [reply lib/improver/utilities/statistical_operations.py]: I'm not sure if this is based on some assumption regarding the input cube. This doesn't seem ideal.

File: lib/improver/utilities/statistical_operations.py [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+        percentile_coordinate, = [coord.name() for coord in
+                                  cube_format.coords()
+                                  if 'percentile' in coord.name()]
+        if percentile_coordinate:
+            probabilities.remove_coord(percentile_coordinate)
+        probabilities.units = 1
+        probabilities.rename(self.output_name)
+        return probabilities
+
+    def percentile_interpolation(self, cube, reference_cube):
```
Comment (gavinevans) [inline lib/improver/utilities/statistical_operations.py]: I think the way that this is working is to find a pair of percentile values and a pair of height values that span the point, in this case a height point, that you want to interpolate to. I think that this is taking an alternative approach to https://github.com/metoppv/improver/blob/master/lib/improver/ensemble_copula_coupling/ensemble_copula_coupling.py#L406, so instead of bookending the height values with a value that is beyond the likely range,  the percentile is repeated (as done after catching the IndexError). Outliers are handled after the interpolation, see the usage of the above_top_band and below_bottom_band variables, instead of prior to the interpolation, as done in https://github.com/metoppv/improver/blob/master/lib/improver/ensemble_copula_coupling/ensemble_copula_coupling.py#L353. 

It would be good to decide if there is an optimum strategy this process, as perhaps naively, I would expect the percentile to probability operation to be a mirror image to the probability to percentile operation, and currently the operations are programmed quite differently.

File: lib/improver/tests/utilities/test_statistical_operations.py [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+
+    def test_attributes(self):
+        """ Test name and units are correctly set """
+        self.assertEqual(self.probability_cube.units, "1")
+        self.assertEqual(self.probability_cube.name(), self.new_name)
+
+    def test_coordinate_collapse(self):
+        """ Test probability cube has no percentile coordinate """
+        with self.assertRaises(CoordinateNotFoundError):
+            percentile_coordinate = \
```
[the remark this answers is not in the document]
Comment (cgsandford) [reply lib/improver/tests/utilities/test_statistical_operations.py]: I'm running this function to check it raises the correct error, but have no reason to check the return value (which is probably None).  What's the preferred convention here?

**Comment by gavinevans** [comment]
@cgsandford For the Codacy message, you can do this:
```python
with self.assertRaises(CoordinateNotFoundError):
    find_percentile_coordinate(self.probability_cube)
```
Therefore, you won't return anything from the function.

**Comment by carolinejjones** [comment]
Well done, this is looking great. One small question. Do we still need the inverse_ordering anymore? We could also probably do we a unit test where we do not set the name (we need some underscores in this it should be probability_of_{cube name)).

File: lib/improver/utilities/statistical_operations.py [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+        pdata = np.full(threshold_heights.shape, np.nan, dtype=float)
+
+        percentile_point_heights_list = []
+        for cube in percentiles_cube.slices(percentile_coordinate):
+            percentile_point_heights_list.append(cube.data.flatten())
+
+        # loop over spatial points in 2D field
+        for index, (percentile_heights, threshold_height) in enumerate(
+                zip(percentile_point_heights_list, threshold_heights)):
+            pdata[index] = 0.01*np.interp(threshold_height, percentile_heights,
```
Comment (gavinevans) [inline lib/improver/utilities/statistical_operations.py]: @cgsandford Depending upon what assumptions are made, this section of code where the cube data is flattened, so that the spatial points can be looped over, could be replaced by:
```python
pdata = np.interp(threshold_cube.data, np.unique(percentiles_cube.data), percentiles)
```
This would mean that there was a single interpolation for all grid points, rather than interpolating separately for each grid point.

File: lib/improver/tests/utilities/test_statistical_operations.py [hunk 11]
```diff
@@ -0,0 +1,278 @@
...
+    reference_array = np.zeros(shape=(16,))
+    reference_array[:4] = 0.
+    reference_array[4:12] = np.linspace(0., 0.875, 8)
+    reference_array[12:] = 1.
+    return reference_array.reshape(4, 4)
+
+
+class Test__init__(IrisTest):
+    """ Test initialisation of the ProbabilitiesFromPercentiles2D class """
+    def setUp(self):
```
Comment (fionaRust) [inline lib/improver/tests/utilities/test_statistical_operations.py]: I think you need tests to show that the `self.inverse_ordering` is set correctly. You probably need two tests, one for True and one for False

File: lib/improver/utilities/statistical_operations.py [hunk 12]
```diff
@@ -0,0 +1,256 @@
...
+            self.percentile_coordinate)
+        self.inverse_ordering = False
+        first_percentile = percentile_slices.next().data
+        for percentile_values in percentile_slices:
+            last_percentile = percentile_values.data
+        if (first_percentile - last_percentile >= 0).all():
+            self.inverse_ordering = True
+
+    def __repr__(self):
+        """Represent the configured plugin instance as a string."""
```
Comment (fionaRust) [inline lib/improver/utilities/statistical_operations.py]: I'm not sure whether you need to print out the `inverse_ordering` in the __repr__ or not. It might be useful.

File: lib/improver/tests/utilities/test_statistical_operations.py [hunk 13]
```diff
@@ -0,0 +1,278 @@
...
+        expected = set_reference_probabilities()
+        expected[np.where(expected <= 0.25)] = 0.
+        expected = 1.0 - expected
+        probability_cube = ProbabilitiesFromPercentiles2D(
+            percentiles_cube).percentile_interpolation(
+                self.orography_cube, percentiles_cube)
+        self.assertArrayAlmostEqual(probability_cube.data, expected)
+
+
+class Test_process(IrisTest):
```
Comment (fionaRust) [inline lib/improver/tests/utilities/test_statistical_operations.py]: I think you need a few more unit tests here, including ones for:
* Test raises the warning
* Test get the right answer if cubes are in different units
* Test what happens when you have leading dimensions on your cube and make sure these don't get lost. See PR #407 for some kind of example of these unit tests and ways of making sure you promote any coords back to dim coords.

File: lib/improver/utilities/statistical_operations.py [hunk 14]
```diff
@@ -0,0 +1,256 @@
...
+                typically x and y.
+        Returns:
+            probabilities (iris.cube.Cube):
+                A 2-dimensional cube of probabilities obtained by interpolating
+                between percentile values.
+        """
+        percentiles = self.percentile_coordinate.points
+        probabilities = self.create_probability_cube(percentiles_cube)
+
+        array_shape = [2] + list(threshold_cube.shape)
```
Comment (fionaRust) [inline lib/improver/utilities/statistical_operations.py]: It is not intuitive why you have a 2 hard-coded here.

File: lib/improver/utilities/statistical_operations.py [hunk 15]
```diff
@@ -0,0 +1,256 @@
...
+        try:
+            probabilities.remove_coord(self.percentile_coordinate)
+        except CoordinateNotFoundError:
+            pass
+
+        probabilities.units = 1
+        probabilities.rename(self.output_name)
+        return probabilities
+
+    def percentile_interpolation(self, threshold_cube, percentiles_cube):
```
Comment (fionaRust) [inline lib/improver/utilities/statistical_operations.py]: I think this whole method needs more comments or explanation throughout it. When you first look at it you do have to think carefully about what each step is doing in order to understand it.

File: lib/improver/utilities/statistical_operations.py [hunk 16]
```diff
@@ -0,0 +1,256 @@
...
+        percentiles = self.percentile_coordinate.points
+        probabilities = self.create_probability_cube(percentiles_cube)
+
+        array_shape = [2] + list(threshold_cube.shape)
+        percentile_bounds = np.full(array_shape, -1, dtype=float)
+        value_bounds = np.full(array_shape, np.nan, dtype=float)
+
+        for index, pslice in enumerate(percentiles_cube.slices_over(
+                self.percentile_coordinate)):
+            # Change to use < and > to force degernate percentile distributions
```
Comment (fionaRust) [inline lib/improver/utilities/statistical_operations.py]: degernate?

Comment (fionaRust) [inline lib/improver/utilities/statistical_operations.py]: We probably need more description in one of the main docstrings about the behaviour for degenerate values in your distribution.

File: lib/improver/tests/utilities/test_statistical_operations.py [hunk 17]
```diff
@@ -0,0 +1,278 @@
...
+
+    def test_coordinate_collapse(self):
+        """ Test probability cube has no percentile coordinate """
+        result = self.pfp_instance.create_probability_cube(
+            self.percentiles_cube)
+        with self.assertRaises(CoordinateNotFoundError):
+            find_percentile_coordinate(result)
+
+
+class Test_percentile_interpolation(IrisTest):
```
Comment (fionaRust) [inline lib/improver/tests/utilities/test_statistical_operations.py]: I'm not sure how far we should go with looking at edge cases, but there could be a few documented in the unit tests. 
One that could be interesting is when you only have one percentile in your input cube.
There's probably others.

File: lib/improver/tests/utilities/test_statistical_operations.py (hunk 11 above)
Comment (bayliffe) [reply lib/improver/tests/utilities/test_statistical_operations.py]: ✅

File: lib/improver/tests/utilities/test_statistical_operations.py (hunk 17 above)
Comment (bayliffe) [reply lib/improver/tests/utilities/test_statistical_operations.py]: I have raised an exception for this to the code and an associated test. We can't do anything useful with a single valued percentile coordinate.

File: lib/improver/utilities/statistical_operations.py (hunk 12 above)
Comment (bayliffe) [reply lib/improver/utilities/statistical_operations.py]: ✅

File: lib/improver/tests/utilities/test_statistical_operations.py (hunk 13 above)
Comment (bayliffe) [reply lib/improver/tests/utilities/test_statistical_operations.py]: * Added warning test to existing 3D test.
* Unit conversion tests added for compatible and incompatible units.
* Added dimension checking as suggested.

File: lib/improver/utilities/statistical_operations.py [hunk 18]
```diff
@@ -0,0 +1,256 @@
...
+            self.inverse_ordering = True
+
+    def __repr__(self):
+        """Represent the configured plugin instance as a string."""
+        result = ('<ProbabilitiesFromPercentiles2D: percentiles_cube: {}, '
+                  'output_name: {}'.format(self.percentiles_cube,
+                                           self.output_name))
+        return result
+
+    def create_probability_cube(self, cube):
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply lib/improver/utilities/statistical_operations.py]: I've created our usual attribute using the inverse_ordering flag to determine if it should be above of below.

File: lib/improver/utilities/statistical_operations.py [hunk 19]
```diff
@@ -0,0 +1,256 @@
...
+
+    def percentile_interpolation(self, threshold_cube, percentiles_cube):
+        """
+        Using a percentiles_cube containing a distinct percentile distribution
+        for each point on a 2-dimensional grid, we can interpolate through each
+        distribution to obtain a probability. The point to which we interpolate
+        is defined by the threshold_cube.
+
+        Args:
+            threshold_cube (iris.cube.Cube):
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply lib/improver/utilities/statistical_operations.py]: ✅

File: lib/improver/utilities/statistical_operations.py (hunk 14 above)
Comment (bayliffe) [reply lib/improver/utilities/statistical_operations.py]: Added a comment.

File: lib/improver/utilities/statistical_operations.py (hunk 16 above)
Comment (bayliffe) [reply lib/improver/utilities/statistical_operations.py]: ✅

File: lib/improver/utilities/statistical_operations.py [hunk 20]
```diff
@@ -0,0 +1,256 @@
...
+            probability_cube (iris.cube.Cube):
+                A cube of probabilities obtained by interpolating between
+                percentile values at the "threshold" level.
+        """
+        cube_slices = self.percentiles_cube.slices(
+            [self.percentile_coordinate, self.percentiles_cube.coord(axis='y'),
+             self.percentiles_cube.coord(axis='x')])
+
+        if threshold_cube.ndim != 2:
+            msg = ('threshold cube has too many ({} > 2) dimensions - slicing '
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply lib/improver/utilities/statistical_operations.py]: I didn't, but it is essentially there to handle orography cubes which often have multiple times or realizations.

File: lib/improver/utilities/statistical_operations.py (hunk 15 above)
Comment (bayliffe) [reply lib/improver/utilities/statistical_operations.py]: I've added a truly enormous doc-string explanation. It's rather a simple process, but it takes quite some explaining.

File: lib/improver/utilities/statistical_operations.py [hunk 21]
```diff
@@ -0,0 +1,427 @@
...
+        Note that the current implementation assumes that in cases of a
+        degenerate percentile distribution, the right most bin in which a
+        threshold value is found is chosen.
+
+        e.g.::
+
+            Percentile: 0 10 20 30 40 50 ...
+            Height (m): 0 0 0 15 30 40 ...
+
+            A height of 0m will be associated with a probabilty of 20%. This is
```
Comment (fionaRust) [inline lib/improver/utilities/statistical_operations.py]: I don't think you want this paragraph to be rendered as code as well, it probably need to go out a level.

Comment (bayliffe) [reply lib/improver/utilities/statistical_operations.py]: ✅

File: lib/improver/utilities/statistical_operations.py [hunk 22]
```diff
@@ -0,0 +1,427 @@
...
+                   [[False, False, False],
+                    [True, True, True],
+                    [True, True, True]]
+
+               The value_bounds array has a leading dimensions with 2 indices
+               to be associated with the lower [0] and upper bounds [1] about
+               the threshold being considered. The [0] index is populated with
+               the values in the slice of percentiles_cube at every True index.
+               The [1] index is populated with the values in the next slice of
+               percentiles_cube.::
```
Comment (fionaRust) [inline lib/improver/utilities/statistical_operations.py]: This does seem to show up as 
`.:` in the documentation followed by the code block.

File: lib/improver/utilities/statistical_operations.py [hunk 23]
```diff
@@ -0,0 +1,427 @@
...
+                      [50, 50, 50]] ]
+
+               Note that where there is no availble +1 index in the
+               percentiles_cube the upper bound is set to be the same as the
+               lower_bound.
+
+            2. When all slices have been interated over, the interpolants are
+               calculated using the threshold values and the values_bounds.
+
+               (threshold_cube.data - lower_bound)/(upper_bound - lower_bound)
```
Comment (fionaRust) [inline lib/improver/utilities/statistical_operations.py]: It might be useful if these lines of maths were code blocks too to make it more readable.

File: lib/improver/tests/utilities/test_statistical_operations.py [hunk 24]
```diff
@@ -0,0 +1,388 @@
...
+        probability_cube = ProbabilitiesFromPercentiles2D(
+            percentiles_cube).percentile_interpolation(
+                self.orography_cube, percentiles_cube)
+        self.assertArrayAlmostEqual(probability_cube.data, expected)
+
+
+class Test_process(IrisTest):
+    """
+    Test top level processing function that calls percentile_interpolation()
+    """
```
Comment (cgsandford) [inline lib/improver/tests/utilities/test_statistical_operations.py]: Rethinking ... I think that throughout this test class we should be comparing the shape of the probability cube with the percentiles cube rather than the orography (threshold) cube, to account for any extra dimensions (as eg at line 335).  Changes needed at lines 289, 299 and 381.

File: lib/improver/tests/utilities/test_statistical_operations.py [hunk 25]
```diff
@@ -0,0 +1,388 @@
...
+
+    def test_coordinate_collapse(self):
+        """ Test probability cube has no percentile coordinate """
+        result = self.pfp_instance.create_probability_cube(
+            self.percentiles_cube)
+        with self.assertRaises(CoordinateNotFoundError):
+            find_percentile_coordinate(result)
+
+
+class Test_percentile_interpolation(IrisTest):
```
Comment (fionaRust) [inline lib/improver/tests/utilities/test_statistical_operations.py]: These tests look good. It might be nice to have one more test showing the result if every single percentile value was the same as an edge case (which is very possible at the moment).

File: lib/improver/utilities/statistical_operations.py [hunk 26]
```diff
@@ -0,0 +1,427 @@
...
+        if threshold_cube.units != self.percentiles_cube.units:
+            threshold_cube.convert_units(self.percentiles_cube.units)
+
+        output_cubes = iris.cube.CubeList()
+        for cube_slice in cube_slices:
+            output_cube = self.percentile_interpolation(threshold_cube,
+                                                        cube_slice)
+            output_cubes.append(output_cube)
+
+        if len(output_cubes) > 1:
```
Comment (fionaRust) [inline lib/improver/utilities/statistical_operations.py]: Do you actually need this if statement? Will merge_cube not work on a cubelist with a single cube in?

File: lib/improver/tests/utilities/test_statistical_operations.py (hunk 13 above)
Comment (fionaRust) [reply lib/improver/tests/utilities/test_statistical_operations.py]: It might be nice to have one process unit test that tests the data is as expected.

File: lib/improver/utilities/statistical_operations.py (hunk 22 above)
Comment (bayliffe) [reply lib/improver/utilities/statistical_operations.py]: ✅

File: lib/improver/utilities/statistical_operations.py (hunk 23 above)
Comment (bayliffe) [reply lib/improver/utilities/statistical_operations.py]: :white

File: lib/improver/tests/utilities/test_statistical_operations.py [hunk 27]
```diff
@@ -0,0 +1,388 @@
...
+
+class Test__init__(IrisTest):
+    """ Test initialisation of the ProbabilitiesFromPercentiles2D class """
+    def setUp(self):
+        """ Set up test cube """
+        self.test_cube = set_up_percentiles_cube()
+        self.new_name = "probability"
+
+    def test_basic(self):
+        """ Test name and default ordering are correctly set """
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply lib/improver/tests/utilities/test_statistical_operations.py]: Changed string.

File: lib/improver/tests/utilities/test_statistical_operations.py [hunk 28]
```diff
@@ -0,0 +1,388 @@
...
+        self.assertEqual(result, expected)
+
+
+class Test_create_probability_cube(IrisTest):
+    """
+    Test creation of new probability cube based on percentile and orography
+    input cubes
+    """
+    def setUp(self):
+        """ Set up a probability cube from percentiles and orography """
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply lib/improver/tests/utilities/test_statistical_operations.py]: ✅

File: lib/improver/tests/utilities/test_statistical_operations.py [hunk 29]
```diff
@@ -0,0 +1,388 @@
...
+        result = self.pfp_instance.create_probability_cube(
+            self.percentiles_cube)
+        self.assertEqual(result.units, "1")
+        self.assertEqual(result.name(), self.new_name)
+        self.assertEqual(result.attributes['relative_to_threshold'], 'below')
+
+    def test_attributes_inverse_ordering(self):
+        """Test relative_to_threshold attribute is suitable for the
+        inverse_ordering case, when it should be 'above'."""
+        self.percentiles_cube.data = np.flipud(self.percentiles_cube.data)
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply lib/improver/tests/utilities/test_statistical_operations.py]: It makes no difference as the setup routine recreates it before each test, so you can do what you like to it within an individual test.

File: lib/improver/tests/utilities/test_statistical_operations.py (hunk 24 above)
Comment (bayliffe) [reply lib/improver/tests/utilities/test_statistical_operations.py]: I've created a reference_cube to use based upon one slice of the percentiles_cube; we obviously have to dispose of the percentiles dimension.

File: lib/improver/utilities/statistical_operations.py [hunk 30]
```diff
@@ -0,0 +1,427 @@
...
+    """
+
+    def __init__(self, percentiles_cube, output_name=None):
+        """
+        Initialise class. Sets an inverse_ordering (bool) switch to true for
+        cases where the percentiled data increases in the opposite sense to the
+        percentile coordinate.
+                e.g.  0th Percentile - Value = 10
+                     10th Percentile - Value = 5
+                     20th Percenitle - Value = 0
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply lib/improver/utilities/statistical_operations.py]: ✅

File: lib/improver/tests/utilities/test_statistical_operations.py [hunk 31]
```diff
@@ -0,0 +1,388 @@
...
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""
+Unit tests for the plugins and functions within statistical_operations.py
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply lib/improver/tests/utilities/test_statistical_operations.py]: Doc string changed. File renamed. Warnings eliminated.

File: lib/improver/tests/utilities/test_statistical_operations.py [hunk 32]
```diff
@@ -0,0 +1,388 @@
...
+def set_up_threshold_cube():
+    """
+    Set up 2D cube with "orography" data on which to threshold percentiles
+    """
+    test_data = 50*np.arange(16).reshape(4, 4)
+    grid_x = DimCoord(np.arange(4), standard_name="projection_x_coordinate",
+                      units="km")
+    grid_y = DimCoord(np.arange(4), standard_name="projection_y_coordinate",
+                      units="km")
+    test_cube = iris.cube.Cube(test_data, long_name="topography", units="m",
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply lib/improver/tests/utilities/test_statistical_operations.py]: ✅

File: lib/improver/tests/utilities/test_statistical_operations.py [hunk 33]
```diff
@@ -0,0 +1,388 @@
...
+        """ Test setting of inverse_order flag using percentiles_cube. In this
+        case the flag should be true as the values associated with the
+        percentiles increase in the opposite direction to the percentiles."""
+        percentiles_cube = self.test_cube.copy(
+            data=np.flipud(self.test_cube.data))
+        pfp_instance = ProbabilitiesFromPercentiles2D(percentiles_cube)
+        self.assertTrue(pfp_instance.inverse_ordering)
+
+    def test_single_percentile(self):
+        """
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply lib/improver/tests/utilities/test_statistical_operations.py]: I've tidies up lots of these doc string triple quotes now.

File: lib/improver/tests/utilities/test_statistical_operations.py [hunk 34]
```diff
@@ -0,0 +1,388 @@
...
+    Test creation of new probability cube based on percentile and orography
+    input cubes
+    """
+    def setUp(self):
+        """ Set up a probability cube from percentiles and orography """
+        self.percentiles_cube = set_up_percentiles_cube()
+        self.percentile_coordinate = find_percentile_coordinate(
+            self.percentiles_cube)
+        self.new_name = "probability"
+        self.pfp_instance = ProbabilitiesFromPercentiles2D(
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply lib/improver/tests/utilities/test_statistical_operations.py]: Agreed. I changed them all to plugin_instance as a compromise.

File: lib/improver/tests/utilities/test_statistical_operations.py [hunk 35]
```diff
@@ -0,0 +1,388 @@
...
+                                                      self.new_name)
+        result = pfp_instance.create_probability_cube(self.percentiles_cube)
+        self.assertEqual(result.attributes['relative_to_threshold'], 'above')
+
+    def test_coordinate_collapse(self):
+        """ Test probability cube has no percentile coordinate """
+        result = self.pfp_instance.create_probability_cube(
+            self.percentiles_cube)
+        with self.assertRaises(CoordinateNotFoundError):
+            find_percentile_coordinate(result)
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply lib/improver/tests/utilities/test_statistical_operations.py]: ✅

File: lib/improver/tests/utilities/test_statistical_operations.py [hunk 36]
```diff
@@ -0,0 +1,388 @@
...
+    def setUp(self):
+        """ Set up orography cube """
+        self.orography_cube = set_up_threshold_cube()
+
+    def test_values(self):
+        """
+        Test that interpolated probabilities at given topography heights are
+        sensible.  Includes out-of-range values (P=0 and P=1).
+        """
+        percentiles_cube = set_up_percentiles_cube()
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply lib/improver/tests/utilities/test_statistical_operations.py]: ✅

File: lib/improver/utilities/statistical_operations.py [hunk 37]
```diff
@@ -0,0 +1,427 @@
...
+        calculated probabilities.
+
+        Args:
+            cube (iris.cube.Cube):
+                Template for the output probability cube.
+        Returns:
+            probability_cube (iris.cube.Cube):
+                A new 2-dimensional probability cube with suitable metadata.
+        """
+        cube_format = next(cube.slices([cube.coord(axis='y'),
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply lib/improver/utilities/statistical_operations.py]: Added notes as we discussed.

File: lib/improver/utilities/statistical_operations.py [hunk 38]
```diff
@@ -0,0 +1,427 @@
...
+        probabilities = cube_format.copy(data=np.full(cube_format.shape,
+                                                      np.nan, dtype=float))
+        try:
+            probabilities.remove_coord(self.percentile_coordinate)
+        except CoordinateNotFoundError:
+            pass
+
+        probabilities.units = 1
+        probabilities.rename(self.output_name)
+        probabilities.attributes['relative_to_threshold'] = 'below'
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply lib/improver/utilities/statistical_operations.py]: I've added `thresholded_using` and then the name of the threshold_cube. This may well need revisiting; should we create a metadata ticket for this?

File: lib/improver/utilities/statistical_operations.py [hunk 39]
```diff
@@ -0,0 +1,427 @@
...
+                A 2-dimensional cube of probabilities obtained by interpolating
+                between percentile values.
+        """
+        percentiles = self.percentile_coordinate.points
+        probabilities = self.create_probability_cube(percentiles_cube)
+
+        # Create array with additional 2 dimensions to contain upper and lower
+        # bounds.
+        array_shape = [2] + list(threshold_cube.shape)
+        percentile_bounds = np.full(array_shape, -1, dtype=float)
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply lib/improver/utilities/statistical_operations.py]: Yeah, I'll leave this be.

File: lib/improver/utilities/statistical_operations.py (hunk 26 above)
Comment (bayliffe) [reply lib/improver/utilities/statistical_operations.py]: It seems to work, so I've taken this out.

File: lib/improver/utilities/statistical_operations.py (hunk 38 above)
[the remark this answers is not in the document]
Comment (gavinevans) [reply lib/improver/utilities/statistical_operations.py]: If this is a new piece of metadata, then yes, as we need to understand the use case for this, and hopefully get some input from Mark.

File: lib/improver/utilities/statistical_operations.py (hunk 21 above)
Comment (fionaRust) [reply lib/improver/utilities/statistical_operations.py]: This still looks weird in the documentation

File: lib/improver/utilities/statistical_operations.py (hunk 37 above)
[the remark this answers is not in the document]
Comment (fionaRust) [reply lib/improver/utilities/statistical_operations.py]: Thank you :)
